File: core/spikes/spike-typed-errors/jvm-harness/src/test/kotlin/TypedErrorTests.kt
// NEN-011 — typed error mapping, Kotlin/JVM tarafı.
// `apple-harness/Tests/SpikeTypedErrorsTests/TypedErrorTests.swift`'in
// (NEN-010) birebir karşılığı.
//
// **Kotlin-özgü isimlendirme bulguları (kanıt kaydına girer, kaynak
// incelemesiyle doğrulandı — bkz. `Main.kt` modül dokümanı):**
// 1. `AppError` → Kotlin'de `AppException` (Error→Exception son ek değişimi,
//    yalnız Kotlin backend'inde; Swift `AppError` kalıyor).
// 2. Düz `uniffi::Enum` varyantları (`Capability`, `ValidationField`) Kotlin'de
//    SCREAMING_SNAKE_CASE üretiliyor (`enum_variant_name` →
//    `to_shouty_snake_case()`), Swift'te ise camelCase (`.localAsr`) idi.
//    Swift zaten PascalCase-error/camelCase-enum asimetrisini bulmuştu; Kotlin
//    üçüncü bir kuralı (SCREAMING_SNAKE_CASE) ekliyor — üç binding boyunca üç
//    farklı isimlendirme kuralı.

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.fail
import uniffi.spike_typed_errors.AppException
import uniffi.spike_typed_errors.Capability
import uniffi.spike_typed_errors.ValidationField
import uniffi.spike_typed_errors.triggerCancelled
import uniffi.spike_typed_errors.triggerCapabilityUnavailable
import uniffi.spike_typed_errors.triggerNetworkError
import uniffi.spike_typed_errors.triggerParseError
import uniffi.spike_typed_errors.triggerValidationError

private const val PRIVATE_PATH = "/Users/alice/Movies/Really Personal/S01E01.mkv"
private const val MEDIA_URL_WITH_TOKEN =
    "https://media.example.com/stream/season1/ep01.mkv?token=eyJhbGciOiJIUzI1NiJ9.super-secret-payload"

class TypedErrorTests {

    /** DoD #1 — her varyant Kotlin tarafında `when` ile, mesaj string'i parse
     *  edilmeden ayrıştırılabiliyor. `else` dalı YOK: derleyici sealed class
     *  exhaustiveness'i zorluyor (Swift'in `default:`'sız `switch`'inin
     *  karşılığı). */
    @Test
    fun everyVariantSwitchesWithoutStringParsing() {
        fun classify(error: AppException): String =
            when (error) {
                is AppException.Parse -> "parse:${error.extension ?: "none"}:${error.line}"
                is AppException.Network -> "network:${error.host}:${error.status}"
                is AppException.Cancelled -> "cancelled"
                is AppException.CapabilityUnavailable -> "capability:${error.capability}"
                is AppException.Validation -> "validation:${error.field}"
            }

        val parse =
            try {
                triggerParseError(PRIVATE_PATH, 7u)
                fail("expected triggerParseError to throw")
            } catch (e: AppException.Parse) {
                e
            }
        assertEquals("parse:mkv:7", classify(parse))

        val network =
            try {
                triggerNetworkError(MEDIA_URL_WITH_TOKEN, 503u)
                fail("expected triggerNetworkError to throw")
            } catch (e: AppException.Network) {
                e
            }
        // media.example.com bu spike'ın allowlist'inde değil.
        assertEquals("network:<redacted-host>:503", classify(network))

        val allowlisted =
            try {
                triggerNetworkError("https://api.opensubtitles.com/v1/x", 200u)
                fail("expected triggerNetworkError to throw")
            } catch (e: AppException.Network) {
                e
            }
        assertEquals("network:api.opensubtitles.com:200", classify(allowlisted))

        try {
            triggerCancelled()
            fail("expected triggerCancelled to throw")
        } catch (e: AppException.Cancelled) {
            assertEquals("cancelled", classify(e))
        }

        val capability =
            try {
                triggerCapabilityUnavailable(Capability.CLOUD_TRANSLATION)
                fail("expected triggerCapabilityUnavailable to throw")
            } catch (e: AppException.CapabilityUnavailable) {
                e
            }
        assertEquals("capability:CLOUD_TRANSLATION", classify(capability))

        val validation =
            try {
                triggerValidationError(ValidationField.MEDIA_DURATION)
                fail("expected triggerValidationError to throw")
            } catch (e: AppException.Validation) {
                e
            }
        assertEquals("validation:MEDIA_DURATION", classify(validation))
    }

    /** DoD #2 — payload'lı bir varyantın Kotlin'in KENDİ varsayılan
     *  `toString()`'i yasaklı desen içermiyor. Rust'ın elle yazılmış `Debug`'ı
     *  burada devrede değil (bkz. `lib.rs` modül dokümanı) — FFI sınırını
     *  gerçekten geçmiş bir hatayı Kotlin'in kendi basımıyla kontrol ediyoruz. */
    @Test
    fun defaultKotlinPrintingHasNoForbiddenPattern() {
        val forbidden = listOf(PRIVATE_PATH, MEDIA_URL_WITH_TOKEN, "media.example.com")

        fun assertSafe(error: AppException, label: String) {
            val printed = error.toString()
            for (pattern in forbidden) {
                assertFalse(
                    printed.contains(pattern),
                    "$label's default Kotlin printing leaked $pattern: $printed")
            }
        }

        try {
            triggerParseError(PRIVATE_PATH, 7u)
        } catch (e: AppException.Parse) {
            assertSafe(e, "parse")
        }

        try {
            triggerNetworkError(MEDIA_URL_WITH_TOKEN, 503u)
        } catch (e: AppException.Network) {
            assertSafe(e, "network")
        }
    }
}

File: core/spikes/spike-typed-errors/apple-harness/Tests/SpikeTypedErrorsTests/TypedErrorTests.swift
import Testing

import SpikeCore

// NEN-010 — typed error mapping across the FFI boundary (M1-core-spike.md).
// swift-testing, not XCTest: this machine has no full Xcode (aynı desen
// spike-async-cancel/apple-harness/Tests ile).
//
// Aynı 8-yasaklı-desen sözlüğü, `nen-domain/tests/guard_redaction.rs` ve
// `spike-typed-errors/src/lib.rs`'nin Rust testleriyle aynı (K23).
private let privatePath = "/Users/alice/Movies/Really Personal/S01E01.mkv"
private let mediaUrlWithToken =
    "https://media.example.com/stream/season1/ep01.mkv?token=eyJhbGciOiJIUzI1NiJ9.super-secret-payload"

@Suite("NEN-010 typed errors")
struct TypedErrorTests {

    /// DoD #1 — her varyant Swift tarafında `switch` ile, mesaj string'i
    /// parse edilmeden ayrıştırılabiliyor. `default:` YOK: derleyici
    /// exhaustiveness'i zorluyor (DoD #3'ün derleme-zamanı yüzü — ayrıca
    /// `scripts/spike-typed-errors.sh`'ın negatif-kontrol adımı çalışma
    /// zamanında değil, tam olarak bu exhaustiveness'i mekanik olarak kırıp
    /// kanıtlıyor).
    @Test("every variant switches without string parsing")
    func everyVariantSwitchesWithoutStringParsing() throws {
        // Not: uniffi 0.32'de `uniffi::Error` varyant adları PascalCase
        // (`.Parse`), sıradan `uniffi::Enum` varyantları ise camelCase
        // (`.localAsr`) — iki farklı derive makrosunun tutarsız isimlendirme
        // kuralı. Bu spike'ın kendisi bu asimetriyi ortaya çıkardı; kanıt
        // kaydında not edilecek.
        func classify(_ error: AppError) -> String {
            switch error {
            case .Parse(let `extension`, let line):
                return "parse:\(`extension` ?? "none"):\(line)"
            case .Network(let host, let status):
                return "network:\(host):\(status)"
            case .Cancelled:
                return "cancelled"
            case .CapabilityUnavailable(let capability):
                return "capability:\(capability)"
            case .Validation(let field):
                return "validation:\(field)"
            }
        }

        #expect(throws: AppError.self) { try triggerParseError(path: privatePath, line: 7) }
        #expect(throws: AppError.self) {
            try triggerNetworkError(url: mediaUrlWithToken, status: 503)
        }
        #expect(throws: AppError.self) { try triggerCancelled() }
        #expect(throws: AppError.self) {
            try triggerCapabilityUnavailable(capability: .localAsr)
        }
        #expect(throws: AppError.self) {
            try triggerValidationError(field: .subtitlePath)
        }

        do {
            try triggerParseError(path: privatePath, line: 7)
            Issue.record("expected triggerParseError to throw")
        } catch let error as AppError {
            #expect(classify(error) == "parse:mkv:7")
        }

        do {
            try triggerNetworkError(url: mediaUrlWithToken, status: 503)
            Issue.record("expected triggerNetworkError to throw")
        } catch let error as AppError {
            // media.example.com is not on the spike's allowlist.
            #expect(classify(error) == "network:<redacted-host>:503")
        }

        do {
            try triggerNetworkError(url: "https://api.opensubtitles.com/v1/x", status: 200)
            Issue.record("expected triggerNetworkError to throw")
        } catch let error as AppError {
            #expect(classify(error) == "network:api.opensubtitles.com:200")
        }

        do {
            try triggerCapabilityUnavailable(capability: .cloudTranslation)
            Issue.record("expected triggerCapabilityUnavailable to throw")
        } catch let error as AppError {
            #expect(classify(error) == "capability:cloudTranslation")
        }

        do {
            try triggerValidationError(field: .mediaDuration)
            Issue.record("expected triggerValidationError to throw")
        } catch let error as AppError {
            #expect(classify(error) == "validation:mediaDuration")
        }
    }

    /// DoD #2 — payload'lı bir varyantın çıktısı yasaklı desen içermiyor.
    /// NEN-006'nın Rust-only kanıtından farkı: burada FFI sınırını GERÇEKTEN
    /// geçmiş bir hatayı, Swift'in KENDİ varsayılan `String(describing:)`
    /// basımıyla kontrol ediyoruz — Rust'ın elle yazılmış `Debug`'ı Swift
    /// tarafında hiç devrede değil (bkz. `lib.rs` modül dokümanı). Ham değer
    /// hiç sınırı geçmediği için burada da yok.
    @Test("caught error's default Swift printing has no forbidden pattern")
    func defaultSwiftPrintingHasNoForbiddenPattern() {
        let forbidden = [privatePath, mediaUrlWithToken, "media.example.com"]

        func assertSafe(_ error: AppError, _ label: String) {
            let printed = String(describing: error)
            for pattern in forbidden {
                #expect(
                    !printed.contains(pattern),
                    "\(label)'s default Swift printing leaked \(pattern): \(printed)"
                )
            }
        }

        do {
            try triggerParseError(path: privatePath, line: 7)
        } catch let error as AppError {
            assertSafe(error, "parse")
        } catch {
            Issue.record("unexpected error type: \(error)")
        }

        do {
            try triggerNetworkError(url: mediaUrlWithToken, status: 503)
        } catch let error as AppError {
            assertSafe(error, "network")
        } catch {
            Issue.record("unexpected error type: \(error)")
        }
    }
}
